### PR TITLE
Api refactoring

### PR DESCRIPTION
**Pull request by danalex97**
This is the first cherry-pick of the massive pull request #212. Most changes relate to the basic API change that moves from sending the whole state each time to sending updates. I also added a modified version of the RaphaelJS front-end so that we will still have a working view of the world.

The view of the world works with localhost and minikube. 

There are problems with the front-end when a minikube game is reloaded. This issues appear as the fix is based on a refactoring that would have been less related to this pull request.

### CODE REVIEW

**Comment by danalex97** [comment]
The coverage decrease comes from the fact that the tests for service were added back after other major changes, so I just kept the test suite version without them
.

**Comment by mikebryant** [comment]
There's some merge comments in there, can we rebase those away?

**Comment by danalex97** [comment]
Ummm, I think both had some conflicts, but nothing major.

**Comment by mikebryant** [comment]
Also can we squash commits:
- f3983a7
- 68c335ef4480ce245a977186acf5e81349d318e3

And squash the merge conflict fixing in:
- ff34dbc

**Comment by mikebryant** [comment]
If you do a rebase, you can fix the conflicts in the commits they happened for, so we get a clean history

Fixing the conflicts in merge commits hides them. We only want clean merge commits in our tree

File: aimmo-game/simulation/world_state.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+        - Changes in map features, i.e. creation and deletion of:
+            * Health points.
+            * Score points.
+            * Pickups.
+            * Obstacles.
+    """
+
+    def __init__(self, game_state):
+        self.game_state = game_state
+        self.ready_to_update = False
```
Comment (mikebryant) [inline aimmo-game/simulation/world_state.py]: I don't think this is needed

File: aimmo-game/service.py [hunk 2]
```diff
@@ -110,28 +57,36 @@ def player_data(player_id):
...
 
-
 def run_game(port):
     global worker_manager
 
     print("Running game...")
     settings = pickle.loads(os.environ['settings'])
+
+    # TODO: this does not work with Kubernates; locally it works
+    # as http://localhost:8000/players/api/games/ is used as default
```
Comment (mikebryant) [inline aimmo-game/service.py]: Lets leave this sort of thing as Github issues, not TODO comments

File: aimmo-game/service.py [hunk 3]
```diff
@@ -110,28 +57,36 @@ def player_data(player_id):
...
 
 if __name__ == '__main__':
     logging.basicConfig(level=logging.INFO)
 
+    # initialize the socket
     socketio.init_app(app, resource=os.environ.get('SOCKETIO_RESOURCE', 'socket.io'))
+
     run_game(int(sys.argv[2]))
+
+    # run the flusk persistent connection
```
Comment (mikebryant) [inline aimmo-game/service.py]: nit: flusk

File: aimmo-game/simulation/action.py [hunk 4]
```diff
@@ -1,6 +1,6 @@
 from logging import getLogger
-from simulation.direction import Direction
-from simulation.event import FailedAttackEvent, FailedMoveEvent, MovedEvent, PerformedAttackEvent, ReceivedAttackEvent
+from direction import Direction
+from event import FailedAttackEvent, FailedMoveEvent, MovedEvent, PerformedAttackEvent, ReceivedAttackEvent
```
Comment (mikebryant) [inline aimmo-game/simulation/action.py]: Please revert these changes, we should be using absolute imports, or proper relative ones.

i.e. the only two correct forms here are:
```
from simulation.direction import Direction
from .direction import Direction
```

File: aimmo-game/simulation/avatar/avatar_manager.py [hunk 5]
```diff
@@ -1,30 +1,61 @@
-from simulation.avatar.avatar_wrapper import AvatarWrapper
-from simulation.avatar.avatar_appearance import AvatarAppearance
-
+from avatar_wrapper import AvatarWrapper
+from avatar_appearance import AvatarAppearance
```
Comment (mikebryant) [inline aimmo-game/simulation/avatar/avatar_manager.py]: Again, fix these imports

File: aimmo-game/simulation/avatar/avatar_manager.py [hunk 6]
```diff
@@ -1,30 +1,61 @@
...
+      it from the scene, it's removed completely.
     """
 
     def __init__(self):
         self.avatars_by_id = {}
+        self.avatars_to_create_by_id = {}
+        self.avatars_to_delete_by_id = {}
 
     def add_avatar(self, player_id, worker_url, location):
+        print("avatar added!!")
```
Comment (mikebryant) [inline aimmo-game/simulation/avatar/avatar_manager.py]: Remove this

File: aimmo-game/simulation/turn_manager.py [hunk 7]
```diff
@@ -3,7 +3,7 @@
 from threading import RLock
 from threading import Thread
 
-from simulation.action import PRIORITIES
+from action import PRIORITIES
```
Comment (mikebryant) [inline aimmo-game/simulation/turn_manager.py]: Again, fix this import

File: aimmo-game/simulation/world_map.py [hunk 8]
```diff
@@ -82,7 +91,7 @@ def _min_max_from_dimensions(cls, height, width):
 
     @classmethod
     def generate_empty_map(cls, height, width, settings):
-        new_settings = map_generator.DEFAULT_LEVEL_SETTINGS.copy()
+        new_settings =DEFAULT_LEVEL_SETTINGS.copy()
```
Comment (mikebryant) [inline aimmo-game/simulation/world_map.py]: nit: style: missing space

File: aimmo-game/simulation/world_map.py [hunk 9]
```diff
@@ -184,7 +203,7 @@ def _update_map(self, num_avatars):
         self._add_pickups(num_avatars)
 
     def _expand(self, num_avatars):
-        LOGGER.info('Expanding map')
+        #LOGGER.info('Expanding map')
```
Comment (mikebryant) [inline aimmo-game/simulation/world_map.py]: Either keep it or remove it, don't comment it

File: players/templates/players/watch.html [hunk 10]
```diff
@@ -16,8 +16,14 @@
...
-<script type="text/javascript" src="/static/js/watch/world-viewer.js"></script>
-<script type="text/javascript" src="/static/js/watch/world-controls.js"></script>
+
+<!-- https://stackoverflow.com/questions/11467873/how-to-append-timestamp-to-the-java-script-file-in-script-tag-url-to-avoid-cac-->
+
+<!-- <script type="text/javascript" src="/static/js/watch/world-viewer.js?v=10"></script> -->
+<!-- <script type="text/javascript" src="/static/js/watch/world-controls.js?v=10"></script> -->
+
+<script>document.write("<script type='text/javascript' src='/static/js/watch/world-viewer.js?v=" + Date.now() + "'><\/script>");</script>
+<script>document.write("<script type='text/javascript' src='/static/js/watch/world-controls.js?v=" + Date.now() + "'><\/script>");</script>
```
Comment (mikebryant) [inline players/templates/players/watch.html]: This is going to be pretty crap for performance. In production I think we handle this automatically for static files, for dev ctrl+f5 works?

File: aimmo-game/simulation/avatar/avatar_manager.py [hunk 11]
```diff
@@ -1,30 +1,61 @@
...
+from avatar_appearance import AvatarAppearance
+import copy
 
 class AvatarManager(object):
     """
-    Stores all game avatars.
+    Stores all game avatars. An avatar can belong to on of the following three lists:
+    - avatars_by_id: If the avatar is on the game.
+    - created_avatars_by_id: If the avatar has ust been created. The avatar is put in
+      this list at first then when the client know about it it's but to avatars_by_id.
```
Comment (mikebryant) [inline aimmo-game/simulation/avatar/avatar_manager.py]: nit: `when the client knows about it it's put into avatars_by_id`

File: aimmo-game/simulation/avatar/avatar_manager.py [hunk 12]
```diff
@@ -1,30 +1,61 @@
...
+        if user_id in self.avatars_by_id:
+            self.avatars_to_delete_by_id[user_id] = copy.deepcopy(self.avatars_by_id[user_id])
+            del self.avatars_by_id[user_id]
 
     @property
     def avatars(self):
         return self.avatars_by_id.viewvalues()
 
+    # Returns the newly created avatars and then puts them in the normal avatars list.
+    @property
```
Comment (mikebryant) [inline aimmo-game/simulation/avatar/avatar_manager.py]: This needs changing - it's a property that performs data modification on a get.. This is confusing, let's change it

File: aimmo-game/simulation/avatar/avatar_manager.py [hunk 13]
```diff
@@ -1,30 +1,61 @@
...
+    def avatars_to_create(self):
+        avatars_to_create_array = list(self.avatars_to_create_by_id.viewvalues())[:]
+        self.avatars_by_id.update(dict(self.avatars_to_create_by_id))
+        self.avatars_to_create_by_id = {}
+
+        return avatars_to_create_array
+
+    # Returns the avatars that need to be removed from the scene.
+    @property
```
Comment (mikebryant) [inline aimmo-game/simulation/avatar/avatar_manager.py]: Again here, we don't want read actions modifying state

File: aimmo-game/simulation/avatar/avatar_manager.py [hunk 14]
```diff
@@ -1,30 +1,61 @@
...
+    def avatars_to_delete(self):
+        avatars_to_delete_array = list(self.avatars_to_delete_by_id.viewvalues())[:]
+        self.avatars_to_delete_by_id = {}
+
+        return avatars_to_delete_array
+
+    # Note: I assume that this will be useful in the future. I have to figure out when
+    # because right now it's just the same as 'avatars'. Otherwise:
+    # TODO: Remove it.
```
Comment (mikebryant) [inline aimmo-game/simulation/avatar/avatar_manager.py]: If it's not referenced anywhere, let's just get rid of it

File: players/templates/players/watch.html (hunk 10 above)
Comment (danalex97) [inline players/templates/players/watch.html]: It does not :( I read how to properly do it, but I thought would be a bit of an overkill as Raphael will not go in production and is just nice to be sure your JS script is updated.

File: aimmo-game/service.py [hunk 15]
```diff
@@ -3,104 +3,51 @@
...
+    world_state.ready_to_update = True
 
 def send_world_update():
-    socketio.emit(
-        'world-update',
-        get_world_state(),
-        broadcast=True,
-    )
-
+    if world_state.ready_to_update:
```
Comment (danalex97) [inline aimmo-game/service.py]: I think all this is useful for the Unity client. I think @ramonfmir knows more.

File: players/templates/players/watch.html (hunk 10 above)
Comment (mikebryant) [inline players/templates/players/watch.html]: Mhm, fair enough, if we're going to strip it out again soon

File: aimmo-game/service.py (hunk 15 above)
Comment (ramonfmir) [inline aimmo-game/service.py]: It is needed because there is a small delay between the socket connection being established and the view being loaded. Since the new API is based on updates, we need to make sure that we don't lose any of the data sent from the back-end. So we use this together with the 'client-ready' request to ensure that the server doesn't start emitting too early.

Comment (ramonfmir) [inline aimmo-game/service.py]: Wait, I understand the confusion. In the 'unity' branch every client has its own world_state. That makes sense with my previous comment. Here we only have one world_state for all of them, in which case the ready_to_update is pointless. We'll remove it from this pull request since it's not strictly related to the API refactoring and add it back later.

**Comment by danalex97** [comment]
I could not squash f3983a7. When I try to run `git rebase -i --preserve-merges` gives more merge conflicts.
